### PR TITLE
`Development`: Add directive to fix types in exercise detail overview list

### PR DESCRIPTION
**Pull request by florian-glombik**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context
When working on [Programming Exercises: Display checkout directories in build plan preview for localCI #8597](https://github.com/ls1intum/Artemis/pull/8597) I had to modify `detail-overview-list.component.hmtl`.

While I think that the generalization approach is very beneficial, I would prefer if we put in some further work and fix the types in that file - as right now there are 105 errors when you open the file!
This makes the code harder to read as IDEs such as IntelliJ will highlight the errors and you have to look up the Enum, the connected type, and then know which attributes are actually present, which seems error prone to me.
It also slows down the IDE quite a bit when opening the file for the first time.

### Description
This PR adds a directive, simplifying the logic that is currently depicted in `detail-overview-list.component.hmtl`, as now the directive will select the appropriate detail component.

Right now, we only use the directive for text-details to keep the scope of the changes manageable and rewieable.
In upcoming PRs components for the other detail types will be added.

* Added a shared component for displaying no data (currently we display `<span>-</span>` when this is the case)
* Using `jhiTranslate` over `artemisTranslate`

<details>
<summary>(Previous approach)</summary>

#### Description
~I added a workaround where~

~1. We add a Pipe that transforms the object to the specific Detail that we have~
~2. We wrap each item in an `@for`-loop that only contains the casted item~
~3. We have fixed the types 🥳 ~

~Unfortunately, we have to add a pipe for each class cast, as I did not find a way to access types in the HTML template - so we can not use a general pipe with generics as we can't pass the type we want to use from the HTML.~

#### Code Review
~I recommend directly comparing `src/main/webapp/app/detail-overview-list/detail-overview-list.component.html` to the version on develop instead of looking at the diff for that file.
Each detail should now be wrapped in an `@for` loop where we cast it to the detail type that we have. Then, we use the caste detail instead of the generic detail. Otherwise, there should be no changes in this file.~
</details>

### Steps for Testing
* 1 instructor

1. Navigate to the details page of a programming exercise (e.g. Java)
2. Verify that the `Package Name` is displayed properly
3. Navigate to the details page of a C programming exercise 
4. Verify that "-" is displayed for `Package Name`
5. Verify that "-" is displayed properly on exam details page in the exercise group table instructor checklist (e.g. if the exercise start is not set, or no points are set)
6. Verify exam summary table headers are translated properly

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots
Code refactoring, the visual representation has not changed
image

Translations in exam checklist are still displayed properly using `jhiTranslate` instead of `artemisTranslate` and `-` is displayed properly for participants
image

Result overview translations are still displayed properly using `jhiTranslate` instead of `artemisTranslate`
image

#### Version without types
image

#### Version **with** directive _(directive support for further components will be added in future PRs)_
The TextDetail is now simplified and we do not get any errors in the file anymore; note that this also means that the IDE will work again for automatic completion and formatting once the directive is used for all sub components.

image

## Summary by CodeRabbit

- **New Features**
  - Introduced `TextDetailComponent` to display text details with a placeholder for missing data.
  - Added `ExerciseDetailDirective` for dynamic component generation based on detail type.
  - Created `NoDataComponent` to handle and display cases where data is absent.

- **Improvements**
  - Replaced text handling in `detail-overview-list` with `jhiExerciseDetail` component for better modularity.
  - Enhanced internationalization using `jhiTranslate` directives in the exam result overview.

- **Bug Fixes**
  - Corrected typo in `alertService` within test files.
  - Added necessary imports and declarations for new components and directives in test files.

### CODE REVIEW

**Comment by florian-glombik** [comment]
Discussed the problem and current approach with @rabeatwork 
> I added a workaround where
> 
> We add a Pipe that transforms the object to the specific Detail that we have
> We wrap each item in an @for-loop that only contains the casted item
> We have fixed the types 🥳
We came to the conclusion that this might fix the typisation issues, but the benefit does not outweigh the added complexity.

We will choose another approach: using Directives and introducing these in multiple PRs instead of one big PR

File: src/main/webapp/app/detail-overview-list/exercise-detail.directive.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        if (!this.isShownDetail()) {
+            return;
+        }
+        this.detail = this.detail as ShownDetail;
+
+        const detailTypeToComponent = {
+            [DetailType.Text]: TextDetailComponent,
+        };
+
+        const detailComponent = detailTypeToComponent[this.detail.type];
```
Comment (florian-glombik) [inline src/main/webapp/app/detail-overview-list/exercise-detail.directive.ts:27]: I will add further components here in the future, this does not happen in this PR to keep the changeset manageable.

File: src/test/javascript/spec/component/detail-overview-list.component.spec.ts [hunk 2]
```diff
@@ -35,11 +36,11 @@ describe('DetailOverviewList', () => {
     let fixture: ComponentFixture<DetailOverviewListComponent>;
     let modalService: NgbModal;
     let modelingService: ModelingExerciseService;
-    let alertServide: AlertService;
+    let alertService: AlertService;
```
Comment (florian-glombik) [inline src/test/javascript/spec/component/detail-overview-list.component.spec.ts:39]: Just fixing a typo here that I noticed when updating the imports

File: src/test/javascript/spec/directive/sidebar-card.directive.spec.ts [hunk 3]
```diff
@@ -64,10 +64,6 @@ describe('SidebarCardDirective', () => {
         jest.resetAllMocks();
     });
 
-    it('should create an instance', () => {
```
Comment (florian-glombik) [inline src/test/javascript/spec/directive/sidebar-card.directive.spec.ts:67]: Seems like the template test for me, I do not think that we need it

The next test checks this aswell anyways

File: src/test/javascript/spec/component/exercise-detail.directive.spec.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    function checkComponentForDetailWasCreated(detailToBeChecked: ShownDetail, expectedComponent: any) {
+        const createComponentSpy = jest.spyOn(component.directive.viewContainerRef, 'createComponent');
+        component.detail = detailToBeChecked;
+        fixture.detectChanges();
+        component.directive.ngOnInit();
+
+        expect(createComponentSpy).toHaveBeenCalledWith(expectedComponent);
+    }
```
Comment (florian-glombik) [inline src/test/javascript/spec/component/exercise-detail.directive.spec.ts:67]: We will add around 15 components, future tests will only need 1 line of code with this method and prevent code duplication

File: src/main/webapp/app/detail-overview-list/components/no-data-component.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { Component } from '@angular/core';
+
+@Component({
+    selector: 'jhi-no-data',
+    template: `<span>-</span>`,
```
[the remark this answers is not in the document]
Comment (florian-glombik) [reply src/main/webapp/app/detail-overview-list/components/no-data-component.ts:5]: The component will be reused in the future and is there to ensure consistency between multiple components

File: src/main/webapp/app/detail-overview-list/exercise-detail.directive.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import { TextDetailComponent } from 'app/detail-overview-list/components/text-detail.component';
+
+@Directive({
+    selector: '[jhiExerciseDetail]',
+    standalone: true,
+})
+export class ExerciseDetailDirective implements OnInit, OnDestroy {
+    @Input() detail: Detail;
+
+    private componentRef: ComponentRef<any>;
```
[the remark this answers is not in the document]
Comment (florian-glombik) [reply src/main/webapp/app/detail-overview-list/exercise-detail.directive.ts:13]: `.instance.detail` does not exist on unknown

File: src/main/webapp/app/detail-overview-list/components/no-data-component.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+    standalone: true,
+})
+export class NoDataComponent {}
```
Comment (JohannesStoehr) [inline src/main/webapp/app/detail-overview-list/components/no-data-component.ts:8]: I think we use `-` for no data in more places in Artemis. Should we move this new component into a common package, so all other components can use it as well?

Comment (florian-glombik) [reply src/main/webapp/app/detail-overview-list/components/no-data-component.ts]: Thanks, I was not aware that we have this solution at multiple places!

I moved the component to the shared folder and re-used it at places where the same solution was used for displaying no data

File: src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.html [hunk 8]
```diff
@@ -79,7 +75,7 @@ <h3>
                     </tbody>
                     <tfoot class="total-footer">
                         <tr>
-                            <th scope="row" colspan="2">{{ 'artemisApp.exam.examSummary.points.total' | artemisTranslate }}</th>
+                            <th scope="row" colspan="2" jhiTranslate="artemisApp.exam.examSummary.points.total"></th>
```
[the remark this answers is not in the document]
Comment (florian-glombik) [reply src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.html:78]: I double checked it and the translation seems to be present in the de and en file

File: src/main/webapp/app/detail-overview-list/exercise-detail.directive.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    ngOnDestroy() {
+        this.componentRef?.destroy();
+    }
+
+    /**
+     * @return false if the detail is a {@link NotShownDetail}
+     */
+    private isShownDetail() {
+        return !!this.detail;
+    }
```
Comment (pzdr7) [inline src/main/webapp/app/detail-overview-list/exercise-detail.directive.ts:43]: Could you add a return type?

**Review by pzdr7 — APPROVED** [review]
Code looks good 👍 One small nitpick

**Review by pzdr7 — APPROVED** [review]
Code 👍

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by Jan-Thurner — APPROVED** [review]
Demonstrated in a testting-session on TS3. Works as expected.

**Review by rstief — APPROVED** [review]
tested on ts3 looks good :) I like the no-data-component

**Review by az108 — APPROVED** [review]
Code LGTM

**Review by rabeatwork — APPROVED** [review]
Maintainer approve

**Pull request merged**